Feature/xyield erc 4626 withdraw

evchip opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
### `xYieldVault.sol`
 - replace updateTotals params with struct
 - add withdrawFrom method for remote withdrawals
 - change updateBalances implementation to support burning or minting tokens
 - add fork test for remoteWithdraw method
<!--- Describe your changes in detail -->

## Motivation and Context
Facilitate same chain withdrawals. Remote withdrawals are partially implemented, but will need bridging capabilities to be fully functional. We are parking this for the time being.
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How Has This Been Tested?
Added fork and unit tests
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Types of changes (remove all unchecked types)

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

-   [x] New feature (non-breaking change which adds functionality)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

-   [x] If my change requires a change to the documentation, I have updated the documentation accordingly, and in either case, have checked this box to attest to my assessment of this requirement with regard to my change.
-   [x] If my change requires additions or updates to any deployment scripts to ensure that the protocol is functional (Makefile, Dockerfile, Forge Script, etc.), I have made these changes, and in either case, have checked this box to attest to my assessment of this requirement with regard to my change.
-   [x] If my change requires additional test coverage, I have created those tests accordingly, and in either case, have checked this box to attest to my assessment of this requirement with regard to my change.

evchip committed:
feat(xYield): initial commit of xYieldVault

establishing simple deposits and withdrawals; add unimplemented functions as todos

evchip committed:
feat(forge modules): install euler vault kit as dependency

evchip committed:
feat(remappings.txt): add euler path and a fix for

bad dependency path on ethereum vault connector

evchip committed:
chore(xYield): forge fmt

evchip committed:
fix(xYieldVault): fix incorrect call in totalAssets

instead of calling yieldProtocol.totalAssets we need to call yieldProtocol.convertToAssets

evchip committed:
test(xYieldForkTest): create 3 tests:

deposit, withdraw, and interest accrual

evchip committed:
feat(xYield): wIP

evchip committed:
fix(xYieldVault): rm T1ERC7683

evchip committed:
feat(xYield): wIP checkpoint

evchip committed:
feat(xYield): wIP checkpoint

evchip committed:
refactor(xYield): wIP cleanup

evchip committed:
refactor(xYield): wip cleanup

evchip committed:
fix(foundry.toml): change solc version back to 0.8.28

to prevent compile issues with incompatible solc versions

evchip committed:
feat(DeployxYieldVault): create scripts to deploy xYield to Arb and Base

evchip committed:
feat(contracts/deployments/xyield/v0.0.1.json): add deployment metadata

evchip committed:
chore(contracts): forge fmt

evchip committed:
chore(workflows/contracts.yml): add env var ARBITRUM_RPC for fork tests

evchip committed:
chore(contracts.yml): add ARBITRUM_BLOCK secret

evchip committed:
feat(xYield): replace updateTotals params with struct;

add withdrawFrom method for remote withdrawals; change updateBalances implementation to support
burning or minting tokens; add fork test for remoteWithdraw method

evchip committed:
refactor(xYieldTest.t.sol): move to tests dir

evchip committed:
refactor(xYieldForkTest): move to test dir

evchip committed:
refactor(xYield): group deposit with other events

evchip committed:
refactor(MockYieldProtocol): move to test dir

evchip committed:
refactor(xYieldVault.sol): change assets to amount

evchip committed:
refactor(xYield tests): fix import paths

evchip committed:
refactor(xYieldVault): prepend underscore to method params for clarity

evchip committed:
refactor(xYieldVault): remove unneeded reentract protection on deposit

evchip committed:
refactor(xYieldVault): remove this from totalAssets

evchip committed:
chore(contracts): merge upstream canary

evchip committed:
fix(xYieldVault): fix incorrect var naming

evchip committed:
feat(xYieldVault): add chainId param to depositFrom

in order to track remote chain

evchip committed:
chore(xYieldVault): forge fmt

evchip committed:
chore(contracts): merge upstream

evchip committed:
refactor(xYield): forge fmt

evchip committed:
refactor(xYield test): fix incorrect params for depositFrom

evchip committed:
refactor(xYieldVault): remove _transferOwnership, redundant

evchip committed:
refactor(xYieldVault): rm _transferOwnership -- redundant

aharvet commented on contracts/src/xYield/xYieldVault.sol:
```diff
@@ -157,26 +161,26 @@ contract xYieldVault is ERC4626, Ownable2Step, ReentrancyGuard, Pausable {
+    // global shares is updated
+    function _withdrawFrom(address owner, uint256 assets) internal {
+        // TODO - deposit into escrow instead of transferring user's underlying to this contract
+        uint256 shares = yieldProtocol.withdraw(assets, address(this), address(this));
```
If I understand well, we plan on potentially integrate with other lending protocols than Morpho and Euler ?
We need to keep in mind that not all of them uses erc 4624 and then they might not return shares here.

evchip replied:
we will only integrate with 4626 vaults

aharvet commented on contracts/src/xYield/xYieldVault.sol:
```diff
@@ -157,26 +161,26 @@ contract xYieldVault is ERC4626, Ownable2Step, ReentrancyGuard, Pausable {
+    function _withdrawFrom(address owner, uint256 assets) internal {
+        // TODO - deposit into escrow instead of transferring user's underlying to this contract
+        uint256 shares = yieldProtocol.withdraw(assets, address(this), address(this));
+        emit WithdrawRemote(_msgSender(), owner, assets, shares);
```
We can use `msg.sender` directly as eip 2771 is dead

evchip replied:
fixed, thanks

aharvet commented on contracts/src/xYield/xYieldVault.sol:
```diff
@@ -17,13 +17,18 @@ contract xYieldVault is ERC4626, Ownable2Step, ReentrancyGuard, Pausable {
+    struct BalanceUpdate {
+        address recipient;
+        uint256 amount;
+        bool isMint;
```
Semantically I would rather use an enum here to clearly stipulate if it's a deposit or a withdrawal. But not a big deal.

evchip replied:
changed, thanks

aharvet commented on contracts/src/xYield/xYieldVault.sol:
```diff
@@ -157,26 +161,26 @@ contract xYieldVault is ERC4626, Ownable2Step, ReentrancyGuard, Pausable {
+    // before
+    // global shares is updated
+    function _withdrawFrom(address owner, uint256 assets) internal {
+        // TODO - deposit into escrow instead of transferring user's underlying to this contract
```
Aren't we supposed to bridge it to caller's chain ?

evchip replied:
correct. we do that via 7683, i.e. locking the user's funds in escrow until the solver fills it on the target chain + provides proof of fill

aharvet replied:
But aren't we supposed to lock those funds on T1ERC7683 by calling `open` function ? or you want to implement custom logic to create an intent and settle after fill directly on xyield contract ?

evchip replied:
yes, we want to open an intent. there are some edge cases around simply calling `open` within `_withdrawFrom`. What happens if the intent is not fulfilled and we need to initiate a refund? we would probably have to add functionality to claim the refund from the xYield contract. I'm still thinking about this, hence why it's marked TODO.

aharvet replied:
Got it

evchip committed:
refactor(xYieldVault.sol): _msgSender() -\> msg.sender

evchip committed:
refactor(xYieldVault): change isMint in BalanceUpdate struct to enum

evchip committed:
chore(contracts): forge fmt

evchip committed:
fix(xYieldVault): fix incorrect formula for totalAssets

total assets should be based on xYield's balance of share tokens; change erc20 transfer calls to
safeTransfer

evchip committed:
test(xYieldForkTest.t.sol): add TODO for mint and redeem tests

evchip committed:
chore(IxYieldVault): add interface for xYield vault

evchip committed:
refactor(xYieldVault): rm unused vars and methods for rebalancing

evchip committed:
chore(contracts): forge fmt

evchip committed:
refactor(contracts): merge upstream

evchip committed:
refactor(contracts): merge upstream

aharvet reviewed (approved):
LGTM
